File: sdks/typescript/src/core/session-nonce.ts
/**
 * DCP v2.0 Session Nonce — anti-splicing defense.
 *
 * A 256-bit random nonce generated once per session and embedded in every
 * artifact. At verification time, the verifier checks that all artifacts
 * in a bundle share the same nonce.
 */

import { randomBytes } from 'crypto';

const SESSION_NONCE_BYTES = 32;
const SESSION_NONCE_HEX_LEN = 64;
const SESSION_NONCE_PATTERN = /^[0-9a-f]{64}$/;

/**
 * Generate a cryptographically random 256-bit session nonce (64 hex chars).
 */
export function generateSessionNonce(): string {
  return randomBytes(SESSION_NONCE_BYTES).toString('hex');
}

/**
 * Validate that a string is a well-formed session nonce.
 */
export function isValidSessionNonce(nonce: string): boolean {
  return (
    typeof nonce === 'string' &&
    nonce.length === SESSION_NONCE_HEX_LEN &&
    SESSION_NONCE_PATTERN.test(nonce)
  );
}

/**
 * Verify that all artifacts in a set share the same session_nonce.
 * Returns { valid, nonce } on success or { valid: false, error } on failure.
 */
export function verifySessionBinding(
  artifacts: Array<{ session_nonce?: string }>,
): { valid: boolean; nonce?: string; error?: string } {
  if (artifacts.length === 0) {
    return { valid: false, error: 'No artifacts to verify' };
  }

  const first = artifacts[0].session_nonce;
  if (!first || !isValidSessionNonce(first)) {
    return { valid: false, error: `Invalid session_nonce in artifact[0]: ${first}` };
  }

  for (let i = 1; i < artifacts.length; i++) {
    const nonce = artifacts[i].session_nonce;
    if (nonce !== first) {
      return {
        valid: false,
        error: `Session nonce mismatch: artifact[0]=${first}, artifact[${i}]=${nonce}`,
      };
    }
  }

  return { valid: true, nonce: first };
}

const DEFAULT_SESSION_DURATIONS: Record<string, number> = {
  routine: 86400,
  standard: 14400,
  elevated: 3600,
  maximum: 900,
};

/**
 * Generate an ISO 8601 session expiry timestamp.
 * @param durationSeconds - how long the session is valid (default: 4 hours)
 */
export function generateSessionExpiry(durationSeconds?: number, tier?: string): string {
  const duration = durationSeconds ?? (tier ? DEFAULT_SESSION_DURATIONS[tier] ?? 14400 : 14400);
  return new Date(Date.now() + duration * 1000).toISOString();
}

/**
 * Check whether a session_expires_at timestamp has passed.
 */
export function isSessionExpired(expiresAt: string): boolean {
  return new Date(expiresAt) < new Date();
}

File: sdks/go/dcp/v2/session_nonce.go
// Package v2 — DCP v2.0 Session Nonce (Go port).

package v2

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"regexp"
	"time"
)

var sessionNoncePattern = regexp.MustCompile(`^[0-9a-f]{64}$`)

// GenerateSessionNonce returns a cryptographically random 256-bit session nonce
// as 64 lowercase hex characters.
func GenerateSessionNonce() string {
	var b [32]byte
	_, _ = rand.Read(b[:])
	return hex.EncodeToString(b[:])
}

// IsValidSessionNonce reports whether s is a well-formed session nonce.
func IsValidSessionNonce(s string) bool {
	return len(s) == 64 && sessionNoncePattern.MatchString(s)
}

// SessionBindingResult is the outcome of VerifySessionBinding.
type SessionBindingResult struct {
	Valid bool
	Nonce string
	Error string
}

// VerifySessionBinding ensures all artifacts share the same session_nonce.
func VerifySessionBinding(artifacts []map[string]interface{}) SessionBindingResult {
	if len(artifacts) == 0 {
		return SessionBindingResult{Error: "No artifacts to verify"}
	}
	first, _ := artifacts[0]["session_nonce"].(string)
	if !IsValidSessionNonce(first) {
		return SessionBindingResult{Error: fmt.Sprintf("Invalid session_nonce in artifact[0]: %q", first)}
	}
	for i, art := range artifacts[1:] {
		nonce, _ := art["session_nonce"].(string)
		if nonce != first {
			return SessionBindingResult{
				Error: fmt.Sprintf("Session nonce mismatch: artifact[0]=%s, artifact[%d]=%s", first, i+1, nonce),
			}
		}
	}
	return SessionBindingResult{Valid: true, Nonce: first}
}

var defaultSessionDurations = map[string]int{
	"routine":  86400,
	"standard": 14400,
	"elevated": 3600,
	"maximum":  900,
}

// GenerateSessionExpiry returns an ISO-8601 UTC timestamp duration seconds in the future.
// If duration is zero, the per-tier default is used (defaulting to 4h when tier is unknown).
func GenerateSessionExpiry(duration int, tier string) string {
	if duration == 0 {
		if d, ok := defaultSessionDurations[tier]; ok {
			duration = d
		} else {
			duration = 14400
		}
	}
	return time.Now().UTC().Add(time.Duration(duration) * time.Second).Format("2006-01-02T15:04:05.000Z")
}

// IsSessionExpired reports whether the ISO-8601 timestamp is in the past.
func IsSessionExpired(expiresAt string) bool {
	t, err := time.Parse(time.RFC3339, expiresAt)
	if err != nil {
		t2, err2 := time.Parse("2006-01-02T15:04:05.000Z", expiresAt)
		if err2 != nil {
			return false
		}
		t = t2
	}
	return t.Before(time.Now().UTC())
}
